src/coord/src/coord.rs

```diff
@@ -5197,6 +5079,126 @@ pub mod fast_path_peek {
+                let collection = compute.collection(*id).unwrap();
+                assert!(collection
+                    .read_capabilities
+                    .frontier()
+                    .less_equal(&read_holds.time));
+                let read_needs = self.read_capability_needs.get_mut(id).unwrap();
+                read_needs.holds.update_iter(Some((read_holds.time, 1)));
+                policy_changes.push((*id, read_needs.policy()));
+            }
+            compute.set_read_policy(policy_changes).await;
+        }
+
+        pub(super) async fn release_read_hold(
+            &mut self,
+            read_holds: ReadHolds<mz_repr::Timestamp>,
+        ) {
+            // We may call this method on an unregistered transaction.
+            let ReadHolds {
+                time,
+                storage_ids,
+                compute_ids,
+                compute_instance,
+            } = read_holds;
+
+            // Update STORAGE read policies.
+            let mut policy_changes = Vec::new();
+            for id in storage_ids.iter() {
+                let read_needs = self.read_capability_needs.get_mut(id).unwrap();
+                read_needs.holds.update_iter(Some((time, -1)));
+                policy_changes.push((*id, read_needs.policy()));
+            }
+            self.dataflow_client
+                .storage()
+                .set_read_policy(policy_changes)
+                .await;
+            // Update COMPUTE read policies
+            let mut policy_changes = Vec::new();
+            for id in compute_ids.iter() {
+                let read_needs = self.read_capability_needs.get_mut(id).unwrap();
+                read_needs.holds.update_iter(Some((time, -1)));
+                policy_changes.push((*id, read_needs.policy()));
+            }
+            self.dataflow_client
+                .compute(compute_instance)
+                .expect("Reference to absent compute instance")
+                .set_read_policy(policy_changes)
+                .await;
+        }
+    }
+}
+
+/// Information about the read capability requirements of a collection.
+struct ReadCapabilityState<T = mz_repr::Timestamp>
+where
+    T: timely::progress::Timestamp,
+{
+    /// The default read policy for the collection when no holds are present.
+    base_policy: ReadPolicy<T>,
+    /// Holds expressed by transactions, that should prevent compaction.
+    holds: MutableAntichain<T>,
+}
+
+impl<T: timely::progress::Timestamp> From<ReadPolicy<T>> for ReadCapabilityState<T> {
+    fn from(base_policy: ReadPolicy<T>) -> Self {
+        Self {
+            base_policy,
+            holds: MutableAntichain::new(),
+        }
+    }
+}
+
+impl<T: timely::progress::Timestamp> ReadCapabilityState<T> {
+    /// Acquires the effective read policy, reflecting both the base policy and any holds.
+    fn policy(&self) -> ReadPolicy<T> {
+        // TODO: This could be "optimized" when `self.holds.frontier` is empty.
```

Comment:
Also if `base_policy` is `ReadPolicy::ValidFrom(Antichain::from_elem(Timestamp::minimum()))`, which will always win.

Reply:
That's true! The comment is there because I initially wrote some code to get the right minimal representation, with one likely common case being "no holds". I gave up and wanted to leave a note.